### PR TITLE
Add type mapping section to Kinesis connector

### PR DESCRIPTION
**Pull request by Jessie212**
## Description

Add type mapping section to the Kinesis connector doc.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by Jessie212** [comment]
@ebyhr Does the type mapping section relate to the [Internal columns](https://trino.io/docs/current/connector/kinesis.html#internal-columns) section at all?

**Comment by hashhar** [comment]
I don't think type mapping applies for this connector since the data type for a given field from a Kinesis message is determined by the table description the user provided (same as the Kafka connector by the way).

So depending on the table description provided by the user the type mapping may differ. Obviously some mappings may not work if the actual type of data in the field and the type in the table description don't match though.

I'm not sure if there's anything to be done here (or in Kafka or Redis which use same concept of a table description supplier).

**Comment by Jessie212** [comment]
> I don't think type mapping applies for this connector since the data type for a given field from a Kinesis message is determined by the table description the user provided (same as the Kafka connector by the way).
> 
> So depending on the table description provided by the user the type mapping may differ. Obviously some mappings may not work if the actual type of data in the field and the type in the table description don't match though.
> 
> I'm not sure if there's anything to be done here (or in Kafka or Redis which use same concept of a table description supplier).

Thanks @hashhar 

cc: @tlblessing @jhlodin

File: docs/src/main/sphinx/connector/kinesis.rst [hunk 1]
```diff
@@ -260,6 +260,70 @@ and if it is a more complex type (JSON array or JSON object) then the JSON itsel
...
+Type mapping
+------------
+
+Because Trino and Kinesis each support types that the other does not, this
+connector :ref:`maps some types <type-mapping-overview>` when
+reading data.
+
+The following table lists Trino data types, which can be used in
+the ``type`` attribute, and the matching field decoders which can be specified in
+the ``dataFormat`` attribute:
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/kinesis.rst:274]: Remove duplicate copy

File: docs/src/main/sphinx/connector/kinesis.rst [hunk 2]
```diff
@@ -260,6 +260,70 @@ and if it is a more complex type (JSON array or JSON object) then the JSON itsel
...
++-------------------------------------+--------------------------------------------------------------------------------+
+| | ``TIME WITH TIME ZONE``           | ``custom-date-time``, ``iso8601``                                              |
++-------------------------------------+--------------------------------------------------------------------------------+
+| | ``TIMESTAMP``                     | ``custom-date-time``, ``iso8601``, ``rfc2822``,                                |
+| |                                   | ``milliseconds-since-epoch``, ``seconds-since-epoch``                          |
++-------------------------------------+--------------------------------------------------------------------------------+
+| | ``TIMESTAMP WITH TIME ZONE``      | ``custom-date-time``, ``iso8601``, ``rfc2822``, ``milliseconds-since-epoch``   |
+| |                                   | ``seconds-since-epoch``                                                        |
++-------------------------------------+--------------------------------------------------------------------------------+
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/kinesis.rst:323]: Convert to standard `list-table` format.

File: docs/src/main/sphinx/connector/kinesis.rst [hunk 3]
```diff
@@ -260,6 +260,70 @@ and if it is a more complex type (JSON array or JSON object) then the JSON itsel
...
+  JSON object.
+* ``formatHint`` - Only for ``custom-date-time``.
+
+The JSON decoder supports multiple field decoders, with ``_default`` being
+used for standard table columns and a number of decoders for date- and
+time-based types.
+
+The following table lists Trino data types, which can be used as in ``type``,
+and matching field decoders, which can be specified via ``dataFormat``
+attribute.
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/kinesis.rst:297]: Suggest rewording for simplicity, something like

> The following table lists Trino data types and their corresponding field decoder, which are specified in `type` and `dataFormat` respectively.

**Review by jhlodin — COMMENTED** [review]
I hope this approach works, so we can have some more consistency between the document streaming connectors with how we talk about table definition files and especially Trino type support. Would like to see what engineers have to say though.

**Review by hashhar — COMMENTED** [review]
this is partial information. Kinesis connector supports all the decoders mentioned in the Kafka docs - not just JSON.

Unless we want to duplicate the content in two places I'd say to leave this page as it is.

Maybe just add a `Type mapping` header under which the existing sentences can go and link to Kafka docs.

**Comment by jhlodin** [comment]
Gotcha, thanks @hashhar . 

We should reduce "to learn about this connector, go read the docs for this other connector" type links as much as possible. Readers don't want to look at a doc for connector X when they're trying to configure connector Y. So, I do think we should duplicate the content into this page, possibly even as a fragment if they're identical. @Jessie212 @mosabua thoughts?

**Comment by hashhar** [comment]
https://github.com/trinodb/trino/pull/14811#issuecomment-1324077698

Let's work on either this or Redis connector first and once we are confident with the approach we can update the other PR.

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 4]
```diff
@@ -0,0 +1,322 @@
+Row decoding
+^^^^^^^^^^^^
+
+For key and message, a decoder is used to map message and key data onto table columns.
+
+The Kinesis connector contains the following decoders:
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/row-decoding.fragment:6]: Need to remove "Kinesis" from this fragment if it's to be reused in Redis and Kafka as well. Suggest a generic "The connector" like we've used in other fragments, and in the list below just "Message is..."

Let's see how that looks. We may also want to break this up into more, smaller fragments specific to each decoder and leave the parent section intact for connector-specific content.

**Review by jhlodin — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+For key and message, a decoder is used to map message and key data onto table
+columns.
+
+The connector contains the following decoders:
+
+* ``raw`` - message is not interpreted; ranges of raw message bytes are mapped to table columns.
+* ``csv`` - message is interpreted as comma separated message, and fields are mapped to table columns.
+* ``json`` - message is parsed as JSON, and JSON fields are mapped to table columns.
+* ``avro`` - message is parsed based on an Avro schema, and Avro fields are mapped to table columns.
+* ``protobuf`` - message is parsed based on a Protobuf schema, and Protobuf fields are mapped to table columns.
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:13]: @nevillelyh Seems `ProtobufDecoderFactory` is added to `DecoderModule` which gets reused by Redsis and Kinesis as well - do you know if it'll work for those as well or is the implementation coupled to Kafka only?

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 6]
```diff
@@ -0,0 +1,323 @@
+Row decoding
+^^^^^^^^^^^^
+
+For key and message, a decoder is used to map message and key data onto table
+columns.
+
+The connector contains the following decoders:
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:7]: Depending on answer from @nevillelyh to my question it might make sense to remove this list since the list can vary per connector.

In such case instead of having a single row-decoding.fragment it might make more sense to have one per decoder like raw-decoder.fragment and so on and then use which ones are applicable to some connector.

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+* ``json`` - message is parsed as JSON, and JSON fields are mapped to table columns.
+* ``avro`` - message is parsed based on an Avro schema, and Avro fields are mapped to table columns.
+* ``protobuf`` - message is parsed based on a Protobuf schema, and Protobuf fields are mapped to table columns.
+
+.. note::
+
+    If no table definition file exists for a table, the ``dummy`` decoder is used,
+    which does not expose any columns.
+
+Raw decoder
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:20]: Assumed copied from Kafka docs, didn't review.

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+* big-endian encoding (for integer types)
+* IEEE 754 format for (for ``DOUBLE``).
+
+Length of decoded byte sequence is implied by the ``dataFormat``.
+
+For ``VARCHAR`` data type a sequence of bytes is interpreted according to UTF-8
+encoding.
+
+CSV decoder
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:89]: Copied from Kafka docs, didn't review.

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
++-------------------------------------+--------------------------------------------------------------------------------+
+| ``BOOLEAN``                         | "true" character sequence maps to ``true``;                                    |
+|                                     | Other character sequences map to ``false``                                     |
++-------------------------------------+--------------------------------------------------------------------------------+
+| ``VARCHAR`` / ``VARCHAR(x)``        | Used as is                                                                     |
++-------------------------------------+--------------------------------------------------------------------------------+
+
+No other types are supported.
+
+JSON decoder
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:123]: Copied from Kafka, didn't review

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+* ``custom-date-time`` - Text based, parses a text field according to Joda format pattern
+                         specified via ``formatHint`` attribute. Format pattern should conform
+                         to https://www.joda.org/joda-time/apidocs/org/joda/time/format/DateTimeFormat.html.
+* ``milliseconds-since-epoch`` - Number-based; interprets a text or number as number of milliseconds since the epoch.
+* ``seconds-since-epoch`` - Number-based; interprets a text or number as number of milliseconds since the epoch.
+
+For ``TIMESTAMP WITH TIME ZONE`` and ``TIME WITH TIME ZONE`` data types, if timezone information is present in decoded value, it will
+be used as Trino value. Otherwise result time zone will be set to ``UTC``.
+
+Avro decoder
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:198]: Copied from Kafka, didn't review

File: docs/src/main/sphinx/connector/row-decoding.fragment [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+* Column is renamed in the new schema:
+  This is equivalent to removing the column and adding a new one, and data created with an older schema
+  produces a *default* value when table is using the new schema.
+
+* Changing type of column in the new schema:
+  If the type coercion is supported by Avro, then the conversion happens. An
+  error is thrown for incompatible types.
+
+Protobuf decoder
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/row-decoding.fragment:255]: Copied from Kafka, didn't review

**Review by hashhar — COMMENTED** [review]
Once these comments are answered and we are aligned on the approach please:

- In first commit copy content from kafka.rst to create row-decoding.fragment and change kafka to use the fragment
- In 2nd commit use fragment in Kinesis
- In 3rd Redis (or separate PR as you see fit)

File: docs/src/main/sphinx/connector/row-decoding.fragment (hunk 5 above)
Comment (nevillelyh) [reply docs/src/main/sphinx/connector/row-decoding.fragment:13]: It should assuming we're reading Protobuf encoded `byte[]` from Redis & Kinesis just like Kafka. Although don't think we have tests covering reading protos from Redis & Kinesis.

Comment (hashhar) [reply docs/src/main/sphinx/connector/row-decoding.fragment:13]: Thanks, I'll create an issue/add some tests - at-least for Redis.

File: docs/src/main/sphinx/connector/row-decoding.fragment (hunk 6 above)
Comment (hashhar) [reply docs/src/main/sphinx/connector/row-decoding.fragment:7]: Let's keep the list as it is.

**Comment by hashhar** [comment]
@Jessie212 I think you can now go forward and address https://github.com/trinodb/trino/pull/14331#discussion_r1044224227 and https://github.com/trinodb/trino/pull/14331#pullrequestreview-1211326117

The approach and content here looks good.

**Comment by Jessie212** [comment]
> Once these comments are answered and we are aligned on the approach please:
> 
>     * In first commit copy content from kafka.rst to create row-decoding.fragment and change kafka to use the fragment
> 
>     * In 2nd commit use fragment in Kinesis
> 
>     * In 3rd Redis (or separate PR as you see fit)

@hashhar I've made the separate commits. I think it's best if I complete the work for Redis in a follow-up.

**Comment by mosabua** [comment]
Ok ... looks like we need to reboot this effort together with the updates in the Kafka connector and the Redis one .. lets sync up about the best plan to get all this done in one sweep @Jessie212 and @jhlodin

**Comment by Jessie212** [comment]
Closing PR. Additional work affects multiple documents and must be split into separate PRs.
